### PR TITLE
[infra][lit-labs/compiler] lit-html tests are now also compiled and tested

### PR DESCRIPTION
**Pull request by AndrewJakubowicz**
Issue: https://github.com/lit/lit/issues/189
RFC: https://github.com/lit/rfcs/pull/21
Full prototype PR: https://github.com/lit/lit/pull/3984

### Context

This is an infrastructure only, and test-only change. No user facing impact.
It does change some lit-html tests, but only slightly such that they can test both uncompiled & compiled templates.

Specifically, any tests that explicitly tested detection of uncompiled templates, need to handle the whole file being compiled.

The compiler also strips markers from comments which required one test to be updated.

### Features

- A new tsconfig in lit-html which only targets the `test/` directory.
- A bespoke TypeScript emitter (in `lit-labs/compiler`) so we can compile the `lit-html/test` directory (and not run into a TS version issue if lit-html was housing the transform).
- A new lit-html test command: `test:dev:compiled`.
- An extra path is handled in the web-test-runner config so the `compiled/` test path can be tested.

### Test plan

This whole PR is a test plan ensuring robust tests for the lit-labs/compiler transform.
It proves that the compiler gracefully handles all the edge cases we test in the lit-html tests.

### Future consideration

Note that in the future when the whole repo is on TS 5.x, we'll be able to use a rollup config to compile the tests instead of a script in the compiler package.

Thank you so much!

### CODE REVIEW

File: packages/labs/compiler/src/test/compile_lit_html_tests.ts
[the remark this answers is not in the document]
Comment (AndrewJakubowicz) [reply packages/labs/compiler/src/test/compile_lit_html_tests.ts:1]: For compatibility with the 2.0 branch and 3.0 branch we cannot do this because lit-html uses TypeScript 4.x.
Thus any usage of the compiler from lit-html results in the transformer crashing because the TypeScript version is wrong.

File: packages/lit-html/src/test/directive-helpers_test.ts [hunk 1]
```diff
@@ -38,6 +38,11 @@ const _$lit_template_1: CompiledTemplate = {
   parts: [],
 };
 
+/**
+ * Use to check if the file has been compiled with @lit-labs/compiler.
+ */
+const isTestFileNotCompiled = isTemplateResult(html``, TemplateResultType.HTML);
```
[the remark this answers is not in the document]
Comment (AndrewJakubowicz) [reply packages/lit-html/src/test/directive-helpers_test.ts:44]: Good point. I guess I can inline some checking logic.

Will change to: 

`html``['_$litType$'] === 1`

File: packages/labs/compiler/src/test/compile_lit_html_tests.ts [hunk 2]
```diff
@@ -0,0 +1,55 @@
...
+
+/**
+ * Set up a TypeScript API program given a tsconfig.json filepath.
+ */
+function programFromTsConfig(tsConfigPath: string): ts.Program {
+  const {config, error} = ts.readConfigFile(tsConfigPath, ts.sys.readFile);
+  if (error) {
+    throw new Error(JSON.stringify(error));
+  }
+  const parsedCommandLine = ts.parseJsonConfigFileContent(
```
[the remark this answers is not in the document]
Comment (AndrewJakubowicz) [reply packages/labs/compiler/src/test/compile_lit_html_tests.ts:22]: oops! Thank you!

File: packages/labs/compiler/src/test/compile_lit_html_tests.ts
[the remark this answers is not in the document]
Comment (augustjk) [reply packages/labs/compiler/src/test/compile_lit_html_tests.ts:1]: Aha, right. I saw that note but totally glanced over the fact that the TS import would be project dependent. Thanks for clarifying!

File: packages/lit-html/src/test/directive-helpers_test.ts (hunk 1 above)
[the remark this answers is not in the document]
Comment (augustjk) [reply packages/lit-html/src/test/directive-helpers_test.ts]: property access on the returned template result? 🤯

**Review by augustjk — APPROVED** [review]
Excellent!

File: packages/lit-html/package.json [hunk 3]
```diff
@@ -344,6 +346,23 @@
         "directives/*.d.ts{,.map}"
       ]
     },
+    "build:compiled:ts:tests": {
+      "#comment": "Compile the test directory with the compiler",
```
Comment (rictic) [inline packages/lit-html/package.json:350]: "the compiler" -> "the lit template compiler" as we've got other tools here that compile like rollup, typescript, etc

fun fact, "compiler" has its roots in the Latin word compilo, a portmanteau of "com" (with, together) and pilo (ram down) that has connotations of both piling things up and of plundering and carrying things off

**Review by rictic — APPROVED** [review]
excellent!

File: packages/lit-html/package.json (hunk 3 above)
Comment (AndrewJakubowicz) [reply packages/lit-html/package.json]: Done! We are plundering `lit-html`'s prepare phase :P
